format-common/src/main/java/io/cdap/plugin/format/charset/fixedlength/FixedLengthCharsetTransformingDecompressorStream.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright © 2020 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.plugin.format.charset.fixedlength;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.apache.hadoop.io.compress.DecompressorStream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.InputStream;
+
+/**
+ * DecompressorStream implementation for the FixedLengthCharsetTransformingDecompressor.
+ * <p>
+ * This can be used to convert input streams containing bytes for fixed length charsets into UTF-8 bytes.
+ */
+public class FixedLengthCharsetTransformingDecompressorStream extends DecompressorStream {
+
+  private final long start;
+  private final long end;
+  private long totalReadBytes = 0;
+
+  public FixedLengthCharsetTransformingDecompressorStream(InputStream in,
+                                                          FixedLengthCharset fixedLengthCharset,
+                                                          long start,
+                                                          long end)
+    throws IOException {
+    super(in, new FixedLengthCharsetTransformingDecompressor(fixedLengthCharset), 4096);
+    in.skip(start);
```

Comment:
Please cast to seekable and use `seek` instead. E.g. on S3 filesystem FSDataInputStream is constructed on top of S3InputStream that do not override default skip, so skip is done in old way - by reading and skipping all the data.

Reply:
@tivv I've amended my commit with your suggestion: https://github.com/cdapio/hydrator-plugins/pull/1255/files#diff-805003eb518a2ad27b1fb8bec65c7fffdbb0bbc9cce75df9edd1d7b2edd161d1R56